scripts/imgtool/keys/ecdsa.py

```diff
@@ -102,37 +102,57 @@ def generate():
     def _get_public(self):
         return self.key.public_key()
 
-    def _build_minimal_ecdsa_privkey(self, der):
+    def _build_minimal_ecdsa_privkey(self, der, format):
+        if (format == serialization.PrivateFormat.OpenSSH):
+            print(os.path.basename(__file__) + \
+                ": Warning: --minimal is supported only for PKCS8 or TraditionalOpenSSL formats")
+            return bytearray(der)
         '''
         Builds a new DER that only includes the EC private key, removing the
         public key that is added as an "optional" BITSTRING.
         '''
-        offset_PUB = 68
         EXCEPTION_TEXT = "Error parsing ecdsa key. Please submit an issue!"
-        if der[offset_PUB] != 0xa1:
-            raise ECDSAUsageError(EXCEPTION_TEXT)
-        len_PUB = der[offset_PUB + 1]
-        b = bytearray(der[:-offset_PUB])
-        offset_SEQ = 29
-        if b[offset_SEQ] != 0x30:
-            raise ECDSAUsageError(EXCEPTION_TEXT)
-        b[offset_SEQ + 1] -= len_PUB
-        offset_OCT_STR = 27
-        if b[offset_OCT_STR] != 0x04:
-            raise ECDSAUsageError(EXCEPTION_TEXT)
-        b[offset_OCT_STR + 1] -= len_PUB
-        if b[0] != 0x30 or b[1] != 0x81:
-            raise ECDSAUsageError(EXCEPTION_TEXT)
-        b[2] -= len_PUB
+        if (format == serialization.PrivateFormat.PKCS8):
+            offset_PUB = 68 # where the context specific TLV starts (tag: 0xA1)
+            if der[offset_PUB] != 0xa1:
+                raise ECDSAUsageError(EXCEPTION_TEXT)
+            len_PUB = der[offset_PUB + 1] + 2 # take into account also the 0xA1 0x44 Type-Length bytes
+            b = bytearray(der[:offset_PUB]) # remove the context specific TLV with the PUB key
+            offset_SEQ = 29
+            if b[offset_SEQ] != 0x30:
+                raise ECDSAUsageError(EXCEPTION_TEXT)
+            b[offset_SEQ + 1] -= len_PUB
+            offset_OCT_STR = 27
+            if b[offset_OCT_STR] != 0x04:
+                raise ECDSAUsageError(EXCEPTION_TEXT)
+            b[offset_OCT_STR + 1] -= len_PUB
+            if b[0] != 0x30 or b[1] != 0x81:
+                raise ECDSAUsageError(EXCEPTION_TEXT)
+            # as b[1] has bit7 set, the length is on b[2]
+            b[2] -= len_PUB
+            if b[2] < 0x80:
+                b = b[0:1] + b[2:]
```

Comment:
I don't think this is needed unless you are using another ASN.1 decoder. The first byte being `0x81` is just telling there is one byte with the length. Although for strict DER this might be more *correct*, but then the Mbed TLS ASN.1 decoder we use doesn't seem to have issues with it the way it is. I will do some testing of this tomorrow or the next day, if it works I have no problem with merging it. PS: you could also have done `del b[1]`!

Reply:
You're right, this is not strictly necessary, the resulting ASN.1 encoding is still compliant, just 1 byte bigger than necessary. But given that I was making changes in this area, I thought of making it strictly adhere to the spec. It shouldn't break any compatibility as both encodings are valid. Will change to del b[1], looks better.